MDS_3/include/CGAL/tetrahedron_soup_to_triangulation_3.h

```diff
@@ -0,0 +1,166 @@
+// File Description :
+//******************************************************************************
+
+#ifndef CGAL_MDS_3_TETRAHEDRON_SOUP_TO_C3T3_H
+#define CGAL_MDS_3_TETRAHEDRON_SOUP_TO_C3T3_H
+
+#include <CGAL/license/MDS_3.h>
+
+#include <CGAL/MDS_3/tet_soup_to_c3t3.h>
+#include <CGAL/internal/MDS_3/MDS_3_helper.h>
+#include <CGAL/boost/graph/named_params_helper.h>
+
+#include <vector>
+#include <array>
+#include <map>
+
+namespace CGAL {
+
+  /** \ingroup PkgMDS3Functions
+  * builds a 3D triangulation from a soup of tetrahedra.
+  *
+  * @tparam TetrahedronRange a model of `Range` whose value type is
+  * a `Tetrahedron_3`
+  * @tparam Triangulation a valid triangulation class that has
+  * a vertex base model of `MeshVertexBase_3`
+  * and a cell base model of `MeshCellBase_3`
+  *
+  * @param tets each element in the range is the geometric description of the
+  * corresponding cell in `tr`
+  * @param tr the 3D triangulation to be built
+  *
+  * @post the output triangulation should be a triangulation of the convex hull of `tets`
+  */
+  template<typename TetrahedronRange, typename Triangulation>
+  void tetrahedron_soup_to_triangulation_3(const TetrahedronRange& tets,
+                                           Triangulation& tr)
+  {
+    typedef Triangulation              Tr;
+    typedef typename Tr::Point         Point;
+
+    std::vector<Point> points;
+    std::vector<std::array<int, 5> > finite_cells;
+    std::map<std::array<int, 3>, typename Tr::Cell::Surface_patch_index> border_facets;
+    std::map<Point, int> p2i;
+
+    CGAL_assertion_code(
+      typename Triangulation::Geom_traits::Orientation_3 orientation =
+      tr.geom_traits().orientation_3_object();
+    );
+
+    for (typename TetrahedronRange::value_type tet : tets)
+    {
+      CGAL_assertion(tet.orientation() == CGAL::POSITIVE);
+      std::array<int, 5> cell;
+
+      for (int i = 0; i < 4; ++i)
+      {
+        const Point& pi = tet[i];
+        if (p2i.find(pi) == p2i.end())
+        {
+          points.push_back(pi);
+          int index = static_cast<int>(points.size() - 1);
+          p2i.insert(std::make_pair(pi, index));
+          cell[i] = index;
+        }
+        else
+          cell[i] = p2i.at(pi);
+      }
+      cell[4] = 1;
+
+      CGAL_assertion(orientation(points[cell[0]],
+        points[cell[1]], points[cell[2]], points[cell[3]]) == CGAL::POSITIVE);
+
+      finite_cells.push_back(cell);
+    }
+
+    CGAL::MDS_3::build_triangulation(tr, points, finite_cells, border_facets);
+
+    CGAL_assertion(CGAL::MDS_3::internal::is_convex(tr));
+  }
+
+  /** \ingroup PkgMDS3Functions
+  * builds a 3D triangulation from a soup of tetrahedra.
+  *
+  * @tparam PointRange  a model of the concept `RandomAccessContainer`
+  * whose value type is the point type
+  * @tparam TetrahedronRange a model of the concept `RandomAccessContainer` whose
+  * value type is a model of the concept `RandomAccessContainer` whose value type is `std::size_t`
```

Comment:
I think that documentation, or the implementation, is wrong, because the function `CGAL::MDS_3::build_triangulation` hardcodes `std::vector` and `std::array<int, 5>`:
```
template<class Tr>
bool build_triangulation(Tr& tr,
                         const std::vector<typename Tr::Point>& points,
                         const std::vector<std::array<int,5> >& finite_cells,
                         const std::map<std::array<int,3>, typename Tr::Cell::Surface_patch_index>& border_facets,
                         std::vector<typename Tr::Vertex_handle>& vertex_handle_vector,
                         const bool verbose = false,
                         bool replace_domain_0 = false)
```

I have tried to replace in `MDS_3/examples/MDS_3/tetrahedron_soup_to_c3t3_example.cpp` a `std::vector` with `boost::container::vector` and I got compilation error about mismatch parameters.

Reply:
Right, I am fixing it.
While writing the documentation, I had hard time finding what is the correct requirement/concept for `std::map`. Is this correct?
```
  *\cgalParamType{a class model of `AssociativeContainer`
  *                    whose key type is model of `RandomAccessContainer`
  *                    and mapped type is `Tr::Cell::Surface_patch_index`}
```

Reply:
I have fixed that issue in [`9c3010b` (#5693)](https://github.com/CGAL/cgal/pull/5693/commits/9c3010b688edb6efe737fdb709ea3d1c4ddd3176), and tested it using this patch:
```patch
diff --git a/TMDS_3/examples/TMDS_3/tetrahedron_soup_to_c3t3_example.cpp b/TMDS_3/examples/TMDS_3/tetrahedron_soup_to_c3t3_example.cpp
index de40d254233..8d95d10f536 100644
--- a/TMDS_3/examples/TMDS_3/tetrahedron_soup_to_c3t3_example.cpp
+++ b/TMDS_3/examples/TMDS_3/tetrahedron_soup_to_c3t3_example.cpp
@@ -9,6 +9,7 @@
 #include <CGAL/IO/File_medit.h>
 
 #include <vector>
+#include <deque>
 #include <unordered_map>
 
 using K = CGAL::Exact_predicates_inexact_constructions_kernel;
@@ -30,9 +31,9 @@ int main(int , char* [])
   DT3 delaunay;
   std::unordered_map<Vertex_handle, int> v2i;
   std::vector<DT3::Point>                points(nbv);
-  std::vector<Tetrahedron_3>             tetrahedra;
-  std::vector<std::array<int, 4> >       tets_by_indices;
-  std::vector< Subdomain_index>          subdomains;
+  std::deque<Tetrahedron_3>             tetrahedra;
+  std::deque<std::vector<int> >         tets_by_indices;
+  std::deque< Subdomain_index>          subdomains;
 
   //insert random points
   CGAL::Random_points_in_cube_3<Point_3> randp(2.);
@@ -44,9 +45,9 @@ int main(int , char* [])
     v2i[v] = i++;
   }
 
-  tetrahedra.reserve(delaunay.number_of_finite_cells());
-  tets_by_indices.reserve(delaunay.number_of_finite_cells());
-  subdomains.reserve(delaunay.number_of_finite_cells());
+  //tetrahedra.reserve(delaunay.number_of_finite_cells());
+  //tets_by_indices.reserve(delaunay.number_of_finite_cells());
+  //subdomains.reserve(delaunay.number_of_finite_cells());
   for (DT3::Cell_handle c : delaunay.finite_cell_handles())
   {
     tetrahedra.push_back(delaunay.tetrahedron(c));
```